File: agents/src/codlock_agents/schemas.py
"""The contract between the orchestrator and the Payment / Fitting agents.

This module is the single source of truth for every payload that crosses an agent
boundary. The orchestrator imports these models (or mirrors them in whatever it is
written in) and codes against them. Nothing here imports the A2A SDK, Gravv, or any
image model on purpose: the contract has to be readable and stable while the
implementations behind it are still being built.

Money is always ``Decimal``, never ``float``. A deposit is a real amount a real person
in Tunisia pays, and binary floating point loses millimes.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class _Base(BaseModel):
    """Strict by default: an unexpected field is a contract drift bug, not a shrug."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)


# --------------------------------------------------------------------------------------
# shared value objects
# --------------------------------------------------------------------------------------


class Money(_Base):
    """An amount in a named currency.

    ``currency`` is the currency the *customer sees* — TND for a Tunisian order. What
    Gravv actually settles in is the Payment Agent's problem, not the orchestrator's,
    and is reported back in :class:`GravvRefs.settlement`.
    """

    amount: Decimal = Field(ge=0, description="Non-negative. Use Decimal, never float.")
    currency: str = Field(default="TND", min_length=3, max_length=3)


class Channel(str, Enum):
    instagram = "instagram"
    whatsapp = "whatsapp"


class CustomerRef(_Base):
    """Who is paying. The Payment Agent needs enough to create a Gravv customer."""

    customer_id: str = Field(description="CODLOCK-side id, e.g. 'C123'.")
    full_name: str
    phone: str = Field(description="E.164 preferred, e.g. '+21620123456'.")
    email: str | None = None
    zone: str | None = Field(
        default=None,
        description="Delivery zone/governorate. Carried for audit; the risk decision "
        "itself is made upstream by the orchestrator's risk tool.",
    )


class ProductRef(_Base):
    """The catalog item the customer is buying, already matched upstream."""

    sku: str
    name: str
    image_url: str = Field(description="Catalog photo, publicly readable.")
    category: str = Field(description="e.g. 'dress', 'shorts', 'shoes'.")
    color: str | None = None
    size: str | None = None


# --------------------------------------------------------------------------------------
# Payment Agent — skill: collect_deposit
# --------------------------------------------------------------------------------------


class DepositStatus(str, Enum):
    not_required = "not_required"
    """Risk scoring returned a zero deposit. No Gravv call was made at all."""

    awaiting_payment = "awaiting_payment"
    """A checkout exists and the customer has not paid yet."""

    paid = "paid"
    expired = "expired"
    failed = "failed"


class GravvRefs(_Base):
    """Gravv-side identifiers, so the backend can reconcile without guessing."""

    seller_account_id: str | None = None
    seller_customer_id: str | None = None
    collection_id: str | None = None
    payment_link_id: str | None = None
    settlement: Money | None = Field(
        default=None,
        description="What Gravv actually moves, if it differs from the displayed "
        "currency (FX applied). None when settlement currency == display currency.",
    )
    environment: Literal["sandbox", "live"] | None = None


class CollectDepositInput(_Base):
    """Orchestrator -> Payment Agent. 'Risk said N, go make it payable.'

    The Payment Agent does **not** decide the amount. ``deposit`` arrives already
    decided by the orchestrator's risk-scoring tool; ``risk_score`` and ``deposit_rate``
    are carried only so the payment record explains itself later.
    """

    order_id: str = Field(description="Idempotency anchor. Two calls with the same "
                                      "order_id must not create two checkouts.")
    seller_id: str
    customer: CustomerRef
    order_total: Money
    deposit: Money = Field(description="Zero is legal and means 'no deposit required'.")
    deposit_rate: float = Field(ge=0, le=1, description="0.20 for 20%. Audit only.")
    risk_score: int = Field(ge=0, le=100, description="Audit only.")
    channel: Channel


class CollectDepositOutput(_Base):
    payment_id: str = Field(description="CODLOCK-side handle for this deposit.")
    order_id: str
    status: DepositStatus
    amount: Money
    checkout_url: str | None = Field(
        default=None, description="The one-tap link the customer opens. None when "
                                  "status is not_required or failed."
    )
    expires_at: datetime | None = None
    gravv: GravvRefs = Field(default_factory=GravvRefs)
    failure_reason: str | None = None


# --------------------------------------------------------------------------------------
# Payment Agent — skill: confirm_payment
# --------------------------------------------------------------------------------------


class ConfirmPaymentInput(_Base):
    payment_id: str
    order_id: str


class ConfirmPaymentOutput(_Base):
    payment_id: str
    order_id: str
    status: DepositStatus
    paid: Money | None = None
    paid_at: datetime | None = None
    failure_reason: str | None = None


# --------------------------------------------------------------------------------------
# Payment Agent — skill: settle_order
# --------------------------------------------------------------------------------------


class OrderOutcome(str, Enum):
    accepted = "accepted"
    refused = "refused"


class Disposition(str, Enum):
    applied_to_total = "applied_to_total"
    """Accepted: the deposit comes off what the customer still owes the courier."""

    retained_for_courier = "retained_for_courier"
    """Refused: the deposit covers the round trip instead of the seller eating it."""

    nothing_to_settle = "nothing_to_settle"
    """No deposit was ever taken (clean returning customer)."""


class SettleOrderInput(_Base):
    order_id: str
    payment_id: str
    outcome: OrderOutcome
    courier_fee: Money | None = Field(
        default=None, description="Required when outcome is refused, to compute whether "
                                  "the deposit actually covered the round trip."
    )


class SettleOrderOutput(_Base):
    order_id: str
    payment_id: str
    disposition: Disposition
    deposit_amount: Money
    remaining_due: Money | None = Field(
        default=None, description="Accepted: what the customer still pays on delivery."
    )
    courier_covered: Money | None = Field(
        default=None, description="Refused: how much of the courier fee the deposit ate."
    )
    seller_shortfall: Money | None = Field(
        default=None, description="Refused: what the seller still loses, if the deposit "
                                  "was smaller than the courier fee. Zero is the win case."
    )


# --------------------------------------------------------------------------------------
# Fitting Agent — skill: generate_preview
# --------------------------------------------------------------------------------------


class PreviewStatus(str, Enum):
    processing = "processing"
    ready = "ready"
    failed = "failed"


class MatchVerdict(str, Enum):
    good_match = "good_match"
    weak_match = "weak_match"
    """The render came out, but the item reads poorly on this customer. The
    orchestrator should offer a catalog alternative — the Fitting Agent deliberately
    has no catalog access of its own, so it reports rather than decides."""


class MatchAssessment(_Base):
    quality: float = Field(ge=0, le=1)
    verdict: MatchVerdict
    reason: str = Field(description="Short human-readable justification, shown to no one "
                                    "but useful at the pitch and in logs.")
    recommend_alternative: bool


class GeneratePreviewInput(_Base):
    request_id: str = Field(description="Idempotency anchor for the render.")
    order_id: str
    customer_photo_url: str
    product: ProductRef


class GeneratePreviewOutput(_Base):
    preview_id: str
    request_id: str
    status: PreviewStatus
    preview_image_url: str | None = Field(
        default=None, description="Set only when status is ready."
    )
    match: MatchAssessment | None = None
    from_cache: bool = Field(
        default=False,
        description="True when served from the pre-rendered fixture set rather than "
        "generated live. Kept explicit so a demo never silently pretends.",
    )
    failure_reason: str | None = None


class GetPreviewInput(_Base):
    """Poll. A render takes 10-30s, which is too long to hold a request open."""

    preview_id: str


__all__ = [
    "Money", "Channel", "CustomerRef", "ProductRef", "GravvRefs",
    "DepositStatus", "CollectDepositInput", "CollectDepositOutput",
    "ConfirmPaymentInput", "ConfirmPaymentOutput",
    "OrderOutcome", "Disposition", "SettleOrderInput", "SettleOrderOutput",
    "PreviewStatus", "MatchVerdict", "MatchAssessment",
    "GeneratePreviewInput", "GeneratePreviewOutput", "GetPreviewInput",
]

File: payment-agent/src/schemas.ts
/**
 * The contract between the orchestrator and the Payment Agent.
 *
 * Single source of truth for every payload crossing the agent boundary. Mirrors
 * `fitting-agent/src/codlock_agents/schemas.py` on the Python side — if you change a
 * shape here, change it there.
 *
 * Money is always a decimal **string**, never a JSON number. TND has three decimal
 * places (millimes) and IEEE floats lose them.
 */

import { z } from 'zod';

/** A decimal amount as a string: "29.800", not 29.8. */
const decimalString = z
  .string()
  .regex(/^\d+(\.\d+)?$/, 'must be a non-negative decimal string, e.g. "29.800"');

export const MoneySchema = z.object({
  amount: decimalString,
  /** What the customer sees. TND for a Tunisian order. */
  currency: z.string().length(3).default('TND'),
});
export type Money = z.infer<typeof MoneySchema>;

export const ChannelSchema = z.enum(['instagram', 'whatsapp']);

export const CustomerRefSchema = z.object({
  customer_id: z.string(),
  full_name: z.string(),
  /** E.164 preferred, e.g. "+21620123456". */
  phone: z.string(),
  email: z.string().nullish(),
  /** Delivery zone. Carried for audit; the risk decision is made upstream. */
  zone: z.string().nullish(),
});

/** Gravv-side identifiers, so the backend can reconcile without guessing. */
export const GravvRefsSchema = z.object({
  seller_account_id: z.string().nullish(),
  seller_customer_id: z.string().nullish(),
  collection_id: z.string().nullish(),
  payment_link_id: z.string().nullish(),
  /**
   * What Gravv actually moves, when it differs from the displayed currency.
   * Gravv holds value in stablecoin and quotes card collections in USD, while
   * orders are priced in TND. Null when the two match.
   */
  settlement: MoneySchema.nullish(),
  environment: z.enum(['sandbox', 'live']).nullish(),
});
export type GravvRefs = z.infer<typeof GravvRefsSchema>;

export const DepositStatusSchema = z.enum([
  /** Risk scoring returned a zero deposit. Gravv was never called. */
  'not_required',
  /** A checkout exists and the customer has not paid yet. */
  'awaiting_payment',
  'paid',
  'expired',
  'failed',
]);
export type DepositStatus = z.infer<typeof DepositStatusSchema>;

// ---------------------------------------------------------------------------------
// skill: collect_deposit
// ---------------------------------------------------------------------------------

/**
 * "Risk said N, go make it payable."
 *
 * The Payment Agent does **not** decide the amount. `deposit` arrives already decided
 * by the orchestrator's risk-scoring tool; `risk_score` and `deposit_rate` are carried
 * only so the payment record explains itself later.
 */
export const CollectDepositInputSchema = z.object({
  /** Idempotency anchor. Two calls with the same order_id must not open two checkouts. */
  order_id: z.string(),
  seller_id: z.string(),
  customer: CustomerRefSchema,
  order_total: MoneySchema,
  /** Zero is legal and means "no deposit required". */
  deposit: MoneySchema,
  /** 0.20 for 20%. Audit only. */
  deposit_rate: z.number().min(0).max(1),
  /** Audit only. */
  risk_score: z.number().int().min(0).max(100),
  channel: ChannelSchema,
});
export type CollectDepositInput = z.infer<typeof CollectDepositInputSchema>;

export const CollectDepositOutputSchema = z.object({
  payment_id: z.string(),
  order_id: z.string(),
  status: DepositStatusSchema,
  amount: MoneySchema,
  /** The one-tap link the customer opens. Null when not_required or failed. */
  checkout_url: z.string().nullable(),
  expires_at: z.string().nullable(),
  gravv: GravvRefsSchema,
  failure_reason: z.string().nullable(),
});
export type CollectDepositOutput = z.infer<typeof CollectDepositOutputSchema>;

// ---------------------------------------------------------------------------------
// skill: confirm_payment
// ---------------------------------------------------------------------------------

export const ConfirmPaymentInputSchema = z.object({
  payment_id: z.string(),
  order_id: z.string(),
});
export type ConfirmPaymentInput = z.infer<typeof ConfirmPaymentInputSchema>;

export const ConfirmPaymentOutputSchema = z.object({
  payment_id: z.string(),
  order_id: z.string(),
  status: DepositStatusSchema,
  paid: MoneySchema.nullable(),
  paid_at: z.string().nullable(),
  failure_reason: z.string().nullable(),
});
export type ConfirmPaymentOutput = z.infer<typeof ConfirmPaymentOutputSchema>;

// ---------------------------------------------------------------------------------
// skill: settle_order
// ---------------------------------------------------------------------------------

export const OrderOutcomeSchema = z.enum(['accepted', 'refused']);
export type OrderOutcome = z.infer<typeof OrderOutcomeSchema>;

export const DispositionSchema = z.enum([
  /** Accepted: the deposit comes off what the customer still owes the courier. */
  'applied_to_total',
  /** Refused: the deposit covers the round trip instead of the seller eating it. */
  'retained_for_courier',
  /** No deposit was ever taken, or it was never actually paid. */
  'nothing_to_settle',
]);
export type Disposition = z.infer<typeof DispositionSchema>;

export const SettleOrderInputSchema = z.object({
  order_id: z.string(),
  payment_id: z.string(),
  outcome: OrderOutcomeSchema,
  /** Required when refused, to compute whether the deposit covered the round trip. */
  courier_fee: MoneySchema.nullish(),
});
export type SettleOrderInput = z.infer<typeof SettleOrderInputSchema>;

export const SettleOrderOutputSchema = z.object({
  order_id: z.string(),
  payment_id: z.string(),
  disposition: DispositionSchema,
  deposit_amount: MoneySchema,
  /** Accepted: what the customer still pays on delivery. */
  remaining_due: MoneySchema.nullable(),
  /** Refused: how much of the courier fee the deposit ate. */
  courier_covered: MoneySchema.nullable(),
  /** Refused: what the seller still loses. Zero is the win case. */
  seller_shortfall: MoneySchema.nullable(),
});
export type SettleOrderOutput = z.infer<typeof SettleOrderOutputSchema>;

/** Every skill this agent answers, with the schema its input must satisfy. */
export const SKILL_INPUTS = {
  collect_deposit: CollectDepositInputSchema,
  confirm_payment: ConfirmPaymentInputSchema,
  settle_order: SettleOrderInputSchema,
} as const;

export type SkillName = keyof typeof SKILL_INPUTS;

export const SKILL_NAMES = Object.keys(SKILL_INPUTS) as SkillName[];
